paho/client.go

```diff
@@ -791,79 +812,154 @@ func (c *Client) Publish(ctx context.Context, p *Publish) (*PublishResponse, err
+	case r := <-resp:
+		switch r.Type {
+		case packets.PUBACK:
+			return PublishResponseFromPuback(r.Content.(*packets.Puback)), nil
+		case packets.PUBREC:
+			return PublishResponseFromPubrec(r.Content.(*packets.Pubrec)), nil
+		case packets.PUBCOMP:
+			return PublishResponseFromPubcomp(r.Content.(*packets.Pubcomp)), nil
+		default:
+			return nil, fmt.Errorf("unexpected response packet type %d", r.Type)
 		}
-	case resp = <-cpCtx.Return:
 	}
+}
 
-	switch pb.QoS {
-	case 1:
-		if resp.Type != packets.PUBACK {
-			return nil, fmt.Errorf("received %d instead of PUBACK", resp.Type)
+func (c *Client) publishQoS12(pb *packets.Publish) (<-chan packets.ControlPacket, error) {
+	c.debug.Println("sending QoS 1 or 2 message")
+	resp := make(chan packets.ControlPacket, 1)
+
+	respInternal := make(chan packets.ControlPacket)
+
+	var err error
+	if pb.PacketID == 0 {
+		pb.PacketID, err = c.MIDs.Request(&CPContext{context.TODO(), respInternal})
+	} else {
+		// ClaimID() is not yet implemented
+		err = fmt.Errorf("publishing with a specific PacketID is not yet implemented")
+	}
+	if err != nil {
+		return nil, err
+	}
+
+	c.workers.Add(1)
+	go func() {
+		defer c.MIDs.Free(pb.PacketID)
+		defer c.workers.Done()
+		defer close(resp)
+
+		if err := c.serverInflight.Acquire(context.Background(), 1); err != nil {
+			go c.error(fmt.Errorf("unexpected failure aquiring inflight semaphore: %w", err))
+			return
 		}
+		defer c.serverInflight.Release(1)
 
-		pr := PublishResponseFromPuback(resp.Content.(*packets.Puback))
-		if pr.ReasonCode >= 0x80 {
-			c.debug.Println("received an error code in Puback:", pr.ReasonCode)
-			return pr, fmt.Errorf("error publishing: %s", resp.Content.(*packets.Puback).Reason())
+		if _, err := pb.WriteTo(c.Conn); err != nil {
+			c.errors.Printf("failed to send PUBLISH for %d: %w", pb.PacketID, err)
+			return
 		}
-		return pr, nil
-	case 2:
-		switch resp.Type {
-		case packets.PUBCOMP:
-			pr := PublishResponseFromPubcomp(resp.Content.(*packets.Pubcomp))
-			return pr, nil
-		case packets.PUBREC:
-			c.debug.Printf("received PUBREC for %s (must have errored)", pb.PacketID)
-			pr := PublishResponseFromPubrec(resp.Content.(*packets.Pubrec))
-			return pr, nil
-		default:
-			return nil, fmt.Errorf("received %d instead of PUBCOMP", resp.Type)
+
+		var receivedPacket packets.ControlPacket
+
+		select {
+		case <-c.stop:
+			c.errors.Printf("PUBLISH operation for %d stopping due to client stop channel", pb.PacketID)
+			return
+		case receivedPacket = <-respInternal:
+		}
+
+		switch pb.QoS {
+		case 1:
+			if receivedPacket.Type != packets.PUBACK {
+				reasonString := fmt.Sprintf(
+					"expected PUBACK for %d, got %s",
+					pb.PacketID,
+					receivedPacket.PacketType(),
+				)
+				c.errors.Printf("disconnecting due to server protocol error. %s", reasonString)
+				go c.Disconnect(&Disconnect{
+					ReasonCode: packets.DisconnectProtocolError,
+					Properties: &DisconnectProperties{ReasonString: reasonString},
+				})
```

Comment:
If we get a protocol error, the client disconnects with the Protocol Error reason code. Based on my understanding of the MQTT spec, this is the correct behavior, but it wasn't being done before.